casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala

```diff
@@ -197,98 +210,61 @@ object InterpreterUtil {
       } yield (preStateHash, postStateHash, processedDeploys, processedSystemDeploys)
     }
 
-  private def computeParentsPostState[F[_]: Sync: BlockStore: Log: Metrics](
+  private def computeParentsPostState[F[_]: Concurrent: BlockStore: Log: Metrics](
       parents: Seq[BlockMessage],
-      dag: BlockDagRepresentation[F],
+      s: CasperSnapshot[F],
       runtimeManager: RuntimeManager[F]
   )(implicit spanF: Span[F]): F[StateHash] =
     spanF.trace(ComputeParentPostStateMetricsSource) {
-      val parentTuplespaces =
-        parents.map(p => p -> ProtoUtil.postStateHash(p))
-
-      parentTuplespaces match {
+      parents.size match {
         // For genesis, use empty trie's root hash
-        case Seq() =>
-          runtimeManager.emptyStateHash.pure
+        case 0 =>
+          RuntimeManager.emptyStateHashFixed.pure[F]
 
-        case Seq((_, parentStateHash)) =>
-          parentStateHash.pure
+        // For single parent, get itd post state hash
+        case 1 =>
+          ProtoUtil.postStateHash(parents.head).pure
 
-        case _ =>
-          findLeastEffortParentForReply(parents, dag).flatMap {
-            case (leastEfforParentIndex, blockHashesToApply) =>
-              replayIntoMergeBlock(
-                parents,
-                dag,
-                runtimeManager,
-                parentTuplespaces(leastEfforParentIndex)._2,
-                blockHashesToApply
-              )
-          }
+        // we might want to take some data from the parent with the most stake,
+        // e.g. bonds map, slashing deploys, bonding deploys.
+        // such system deploys are not mergeable, so take them from one of the parents.
+        case _ => {
+          implicit val r = runtimeManager
+          val tipsHashes = parents.map(_.blockHash)
+          val lfbHash    = s.lastFinalizedBlock
```

Comment:
I just want to say that I still think it is wrong to introduce last finalized block here for state merging.  I think you should check if the base state is right when you checkout 0.67 synchrony constraint.

Maybe it would be better to think as @tgrospic mentioned the concept -> deploy chain before and stop thinking base state.

Reply:
@zsluedem There is no common base state in this implementation. LFB here is used to determine the scope of merging.
Deploy chains should be calculated somehow - you might think about current code as one of possible implementations of how to calculate these event chains.
1. You have DAG with tips
2. You have a scope defined, which is subject to change (deploys within scope can be rejected). This scope is - from tips, down to finalised block.
3. Then you need to split this scope in set of deploy chains - this is what is done in https://github.com/rchain/rchain/blob/5dac4d19ffbc84e85608dee780419b55009134d8/casper/src/main/scala/coop/rchain/casper/blocks/merger/CasperDagMerger.scala#L72-L73
All tips should be descendants of finalized block as per block creator (lets call it LFBc) (** *actually this statement might need more thinking - here I agree, with 0.67 it might require modification. I'm not sure if it is guaranteed that all tips have LFB as an ancestor, thanks for heads up*), so DAG is traversed and subdag connecting all tips to LFBc defines the scope of merge.
This particular way of splitting aims having one of tips as the mains state. This should minimise number of events to merge, because many of events in the scope are already inside tip state.
4. Then you have set of deploy chains that you can merge. Each deploy chain is defined by `start state` (post state of tip of the DAG), `end state` (the very bottom block which is starting point of event chain **Note: this is NOT LFB**) and all events that happened to transition start to end.
https://github.com/rchain/rchain/blob/cf7943ef955474b0f5b53c3530ee8a5625eadffc/rspace/src/main/scala/coop/rchain/rspace/merger/StateMerger.scala#L24-L34

PTAL at remark under ** in item 3

Reply:
This actually returns us again to older question - how to define the scope of merging. As synchrony constraint is not connected to safety oracle, it's not that straightforward. Node A might have last finalised block that is not connected to all tips when it is about to create block.

Reply:
Please add this info on a new issue so we can extend discussion.